### ISSUE TITLE
Suggest `uN::midpoint` over overflow prone `(a + b) / 2` expressions

### ISSUE THREAD

**Issue by Urgau** [rust-lang/rust-clippy#13849]
### What it does

This lint is would suggest rewrite the not quite correct `(a + b) / 2` expression to the ([recently stabilized](https://github.com/rust-lang/rust/pull/131784)) [`uN::midpoint`/`iN::midpoint`/`fN::midpoint`](https://doc.rust-lang.org/nightly/std/primitive.u64.html#method.midpoint) functions, in order to prevent overflow of `a + b`, which the `midpoint` functions handle appropriately.

### Advantage

- Eliminate a potential source of overflow
  - [Java’s binary search implementation](https://research.googleblog.com/2006/06/extra-extra-read-all-about-it-nearly.html) and [Mozilla](https://bugzilla.mozilla.org/show_bug.cgi?id=917841) had the overflow bug
- Bring awareness to the surprisingly subtleness of midpoint 

### Drawbacks

- Introduce a function call *(but at the same the implementation is very small and the functions are inlined 🤷)*

### Example

```rust
let a = 10_i32;
let b = 10_000_i32;

let mid = (a + b) / 2;
```

Should be written as:

```rust
let a = 10_i32;
let b = 10_000_i32;

let mid = a.midpoint(b);
```

**Comment by samueltardieu** [comment]
Interesting

@rustbot claim
